tripal/src/Services/TripalEntityLookup.php

```diff
@@ -0,0 +1,270 @@
+      $renderable_item = [
+        '#markup' => $displayed_string,
+      ];
+    }
+
+    return $renderable_item;
+  }
+
+  /**
+   * Retrieve a Drupal entity ID for a record in a given bundle given the bundle's CV term.
+   *
+   * @param int $record_id
+   *   The primary key value for the requested record
+   * @param string $termIdSpace
+   *   The bundle's CV Term namespace e.g. for gene "SO"
+   * @param string $termAccession
+   *   The bundle's CV term accession e.g. for gene "0000704"
+   * @param string $entity_type
+   *   The type of entity, only 'tripal_entity' is supported.
+   *
+   * @return int|null
+   *   The Drupal entity ID, or null if no match found.
+   */
+  public function getEntityId($record_id, $termIdSpace, $termAccession, $entity_type = 'tripal_entity') {
+
+    // Catch invalid entity type
+    if ($entity_type != 'tripal_entity') {
+      throw new \Exception("Invalid entity type \"$entity_type\". getEntityId() only supports the entity type \"tripal_entity\"");
+    }
+
+    // Perform the lookup steps
+    $entity_id = NULL;
+    $bundle_id = $this->getBundleFromCvTerm($termIdSpace, $termAccession, $entity_type);
+    if ($bundle_id) {
+      $base_table = $this->getBundleBaseTable($bundle_id, $entity_type);
+      if ($base_table) {
+        $entity_ids = $this->getEntityIdFromRecordId($base_table, $record_id, $bundle_id, $entity_type);
+        if ($entity_ids) {
+          // Here we are just returning the first hit, e.g. analysis published as both
+          // analysis and genome assembly. Ideally this will be prevented from happening.
+          $entity_id = $entity_ids[0];
+        }
+      }
+    }
+
+    return $entity_id;
+  }
+
+  /**
+   * Retrieve a Tripal bundle id based on its CV term.
+   *
+   * @param string $termIdSpace
+   *   The bundle's CV Term namespace e.g. "NCIT"
+   * @param string $termAccession
+   *   The bundle's CV term accession e.g. "C47954"
+   * @param string $entity_type
+   *   The type of entity, only 'tripal_entity' is supported.
+   *
+   * @return string|null
+   *   The bundle id, or null if no match found.
+   */
+  protected function getBundleFromCvTerm($termIdSpace, $termAccession, $entity_type) {
+    $bundle_id = NULL;
+    $bundle_manager = \Drupal::service('entity_type.bundle.info');
+    $bundle_list = $bundle_manager->getBundleInfo($entity_type);
+    foreach ($bundle_list as $id => $properties) {
+      // Get each bundle's CV term
+      $bundle_info = \Drupal::entityTypeManager()->getStorage('tripal_entity_type')->load($id);
+      $bundleIdSpace = $bundle_info->getTermIdSpace();
+      $bundleAccession = $bundle_info->getTermAccession();
+
+      // Find the bundle where the term values match
+      if (($termIdSpace == $bundleIdSpace) and ($termAccession == $bundleAccession)) {
+        $bundle_id = $id;
+        break;
+      }
+    }
+    return $bundle_id;
+  }
+
+  /**
+   * Retrieve the base table for a given bundle given the bundle's CV term.
```

Comment:
Since this function looks up the base table given the bundle, it doesn't actually matter how the bundle was retrieved.

```suggestion
   * Retrieve the base table for a given bundle.
```